Validation/MuonHits/src/DTSimHitMatcher.cc

```diff
@@ -0,0 +1,374 @@
+#include "Validation/MuonHits/interface/DTSimHitMatcher.h"
+
+using namespace std;
+
+DTSimHitMatcher::DTSimHitMatcher(const edm::ParameterSet& ps,
+                                 edm::ConsumesCollector&& iC)
+    : MuonSimHitMatcher(ps, std::move(iC)) {
+  simHitPSet_ = ps.getParameterSet("dtSimHit");
+  verbose_ = simHitPSet_.getParameter<int>("verbose");
+  simMuOnly_ = simHitPSet_.getParameter<bool>("simMuOnly");
+  discardEleHits_ = simHitPSet_.getParameter<bool>("discardEleHits");
+
+  simHitInput_ = iC.consumes<edm::PSimHitContainer>(
+      simHitPSet_.getParameter<edm::InputTag>("inputTag"));
+}
+
+/// initialize the event
+void DTSimHitMatcher::init(const edm::Event& iEvent,
+                           const edm::EventSetup& iSetup) {
+  iSetup.get<MuonGeometryRecord>().get(dt_geom_);
+  if (dt_geom_.isValid()) {
+    geometry_ = &*dt_geom_;
+  } else {
+    hasGeometry_ = false;
+    edm::LogWarning("DTSimHitMatcher")
+        << "+++ Info: DT geometry is unavailable. +++\n";
+  }
+  MuonSimHitMatcher::init(iEvent, iSetup);
+}
+
+/// do the matching
+void DTSimHitMatcher::match(const SimTrack& track, const SimVertex& vertex) {
+  MuonSimHitMatcher::match(track, vertex);
+
+  if (hasGeometry_) {
+    matchSimHitsToSimTrack(track_ids_, simHits_);
+
+    if (verbose_) {
+      edm::LogInfo("DTSimHitMatcher")
+          << "nTrackIds " << track_ids_.size() << " nSelectedDTSimHits "
+          << hits_.size() << endl;
+      edm::LogInfo("DTSimHitMatcher")
+          << "detids DT " << detIds(0).size() << endl;
+
+      const auto& dt_det_ids = detIds(0);
+      for (const auto& id : dt_det_ids) {
+        const auto& dt_simhits = MuonSimHitMatcher::hitsInDetId(id);
+        const auto& dt_simhits_gp = simHitsMeanPosition(dt_simhits);
+        edm::LogInfo("DTSimHitMatcher")
+            << "DTWireId " << DTWireId(id) << ": nHits " << dt_simhits.size()
+            << " eta " << dt_simhits_gp.eta() << " phi " << dt_simhits_gp.phi()
+            << " nCh " << chamber_to_hits_[id].size() << endl;
+      }
+    }
+  }
+}
+
+void DTSimHitMatcher::matchSimHitsToSimTrack(
+    std::vector<unsigned int> track_ids, const edm::PSimHitContainer& dt_hits) {
```

Comment:
@dildick why is track_ids passed by value and not by reference, may eb as a const, as it looks you do not want to change it? Not sure how big it is...